Create a new ibis post about the backends. (#89)
Quansight-Labs/quansight-labs-site: posts/2020/07/the-ibis-backends.md

```python
    import ibis.omniscidb, dask, intake, sqlalchemy, pandas, pyarrow as arrow, altair, h5py as hdf5
```

<!-- END_TEASER -->

<!--    

 
-->

# `ibis` as a generalized query tool for different backends

> In [our most recent `ibis` post] we look at querying & retrieving data using a familiar `pandas`-like interface.

Reviewer: missing link

The design of quite a few high-level tools in the scientific Python world can be tracked back to the holistic `blaze` ecosystem that offered ways _**store, describe, query, and process**_ data.
`blaze` had ambitious goals and is now defunct, but its influences resonate throughout
the scientific python community development in the success of projects the focus specific features of a data ecosystem like:

* `dask` processing data.
* `intake` for describing data.
* `ibis` for querying data.

Throughout the rest of this document we'll highlight the ability of `ibis` to generically prescribe
query expressions across different data storage systems.

### The design of [`ibis` backends][backends]

Currently, `ibis` supports __>10__ backends.

    >>> dir(ibis)
    [...HDFS...WebHDFS...bigquery...clickhouse...hdf5...impala...omniscidb...pandas...pyspark...spark...sql...sqlite...]

A backend takes an `ibis` query expression and applies computation, _and the query is independent of the computation_.
A backend implementation, that can be queried with `ibis`, has one of the three following architectures.

> 1. Direct execution backends - `pandas and hdf5`.

Reviewer: ```suggestion
1. Direct execution backends - `pandas` and `hdf5`.
```

    <tr>
      <th>2017</th>
      <td></td>
      <td>1</td>
      <td>21</td>
      <td>15</td>
      <td>49</td>
      <td>10</td>
      <td>15</td>
      <td></td>
      <td></td>
      <td>8</td>
      <td></td>
      <td>10</td>
    </tr>
    <tr>
      <th>2018</th>
      <td>31</td>
      <td></td>
      <td>10</td>
      <td>71</td>
      <td>35</td>
      <td>8</td>
      <td>17</td>
      <td></td>
      <td></td>
      <td>9</td>
      <td>2</td>
      <td>2</td>
    </tr>
    <tr>
      <th>2019</th>
      <td>33</td>
      <td>22</td>
      <td>17</td>
      <td>12</td>
      <td>32</td>
      <td>1</td>
      <td>4</td>
      <td></td>
      <td>7</td>
      <td>1</td>
      <td>2</td>
      <td>5</td>
    </tr>
    <tr>
      <th>2020</th>
      <td>38</td>
      <td>3</td>
      <td>4</td>
      <td>2</td>
      <td>4</td>
      <td>1</td>
      <td>2</td>
      <td>1</td>
      <td>3</td>
      <td>4</td>
      <td>4</td>
      <td></td>
    </tr>
  </tbody>ibis direct execution backends like pandas and hdf5 operate on conventional in-memory python objects. pandas is the gold standard for structured data in python, and inspires the API for ibis.

</table>
</div>

> 

Reviewer: ```suggestion
<br>
```
adding an extra line break for visual appeal

> > We note an early focus `ibis.sqlite, sqlalchemy and ibis.impala`. 

Reviewer: ```suggestion
> We note an early focus `ibis.sqlite`, `sqlalchemy` and `ibis.impala`. 
```

Later, work began on the `pandas` backend rounding out the three different types of backgrounds.
From this point, improvements were made to these key backends as `ibis.clickhouse`, `ibis.spark` and `ibis.postgres`. 
For the past 3 years, Quansight, in partnership with OmniSci, added the `ibis.omniscidb`
string generating backend. Further, our responsibilities have expanded
to support `ibis` as community maintainers through Quansight Labs. 
This collaboration introduced geospatial functionality to `ibis` for several backends.

### Currently, there is an ongoing effort to add sqlserver backends

> There are on going efforts to introduce [SQL Server][sql-server] support.

Reviewer: ```suggestion
There are on going efforts to introduce [SQL Server][sql-server] support.  
```

Reviewer: missing link?

What other backends would you like to see for `ibis`?
Maybe our community could benefit from a `dask` backend or `altair` backend?

## `ibis` direct execution backends

`ibis` direct execution backends like `pandas` and `hdf5` operate on conventional in-memory python objects.
`pandas` is the gold standard for structured data in python, and inspires the API for `ibis`.

```python
    pd = ibis.pandas.connect({'A': pandas.util.testing.makeDataFrame()})
```

> `pd` is an `ibis` backend based off `pandibis direct execution backends like pandas and hdf5 operate on conventional in-memory python objects. pandas is the gold standard for structured data in python, and inspires the API for ibis.

Reviewer: ```suggestion
`pd` is an `ibis` backend based off `pandas.DataFrame` objects.
```

Reviewer: seems like this was errant paste. I think I may have fixed it?

> 

Reviewer: ```suggestion
```

> as.DataFrame` objects.

Reviewer: ```suggestion
```

```python
    expression = pd.table('A').head()
```

`expression` is an `ibis` query, that has `expression.compile` and `expression.execute` methods.
We'll recognize the __execute__ method when we return `pandas.DataFrame`s from `ibis` expression.
The __compile__ method does not trigger any computation, rather it constructs an intermediate form
that is interpretted by a backend.

```python
    >>> assert isinstance(expression.compile(), ibis.expr.types.TableExpr)
```

In the case of direction execution backends, the `expression` compiles to an the original `ibis` 
expression.  And the computation is carried out based on a set of recipes defined in `ibis`.

In general, we would typically do this work directly in `pandas`, however this work is
practical in mocking tests for expressions independent of backends.

> Learn more about the [HDF5 direct execution backend in the `ibis` tests][test-hdf5].

## `ibis` expression generating backends.

```python
    db = ibis.sqlite.connect('lahmansbaseballdb.sqlite')
    expression = db.table('halloffame').head()
```

> Expression generating backends operate on [SQL] databases that interoperate with `sqlalchemy`.

Reviewer: link?

```python
    >>> assert isinstance(expression.compile(), sqlalchemy.sql.Select)
```

In the case of expression generating backends, the intermediate representation is a `sqlalchemy` object.
`sqlalchemy` is _The Database Toolkit for Python_, and `ibis` leverages it compatability

> with traditional [SQL] databases.

Reviewer: link?

    

## `ibis` string generating backends.

```bash
pip install --upgrade ibis-framework[omniscidb]

> conda install -c conda-forge ibis-framework # install all the backends!

Reviewer: insert "OR" line in between.

```

String generating backends allow `ibis` to interface with big data systems that manage 
their own computation. For example, we may connect to an example `omnisci` database.
    

```python
    import ibis.omniscidb
    omnisci = ibis.omniscidb.connect(host='metis.omnisci.com', user='demouser', password='HyperInteractive', port=443, database='omnisci', protocol='https')
```
    
`omnisci` is described as a string generating backend because the intermediate representation of the
query is a flavor of SQL.

```python
    expression = omnisci.table('upstream_reservoir').head()
```

    

> A string generating expression compiles to `ibis.omniscidb` flavored [SQL], while `ibis.bigquery` may have a different string representation.

Reviewer: link to SQL?

```python
    >>> expression.compile()
    'SELECT *\nFROM upstream_reservoir\nLIMIT 5'
```
    

> Major credit goes to [@xmnlab] in his heroic PR to introduce `ibis.omniscidb` into `ibis`. You can watch
the drama play out in this [Github Issue][omnisci-pr]. If you'd like to learn more about [OmniSci] and
`ibis.omniscidb` checkout the following links.
> * [OmniSci][omnisci]
> * [Quansight Labs- Ibis: Python data analysis productivity framework][labs-post]

## Conclusion

We'd like to thank the maintainers of the `ibis` for
their and effort in supporting the `ibis` community.

> [sql-server]: #

Reviewer: ah I now see where the links are... so many of my above links comments may be invalid. But it seems that some of these are still empty.

This example focused on the fluent API that `ibis` provides to query structure from a SQLite database, a single backend.
In this post, we'll explore `ibis`'s ability to answer questions about data using to different `ibis` backends.

## `ibis` in the scientific Python ecosystem.

Before we continue into the technical nitty-gritty of `ibis`, we'll consider `ibis` in the greater historical context of the scientific Python ecosystem.

The design of quite a few high-level tools in the scientific Python world can be tracked back to the holistic `blaze` ecosystem that offered ways _**store, describe, query, and process**_ data.
`blaze` had ambitious goals and is now defunct, but its influences resonate throughout
the scientific python community development in the success of projects the focus specific features of a data ecosystem like:

* `dask` processing data.
* `intake` for describing data.
* `ibis` for querying data.

Throughout the rest of this document we'll highlight the ability of `ibis` to generically prescribe
query expressions across different data storage systems.

> ### The design of [`ibis` backends][backends].

Author: the title doesnt make sense?

> ### The design of [`ibis` backends][backends].

Author: i think this is a "links" comment

      <td></td>
      <td>10</td>
    </tr>
    <tr>
      <th>2018</th>
      <td>31</td>
      <td></td>
      <td>10</td>
      <td>71</td>
      <td>35</td>
      <td>8</td>
      <td>17</td>
      <td></td>
      <td></td>
      <td>9</td>
      <td>2</td>
      <td>2</td>
    </tr>
    <tr>
      <th>2019</th>
      <td>33</td>
      <td>22</td>
      <td>17</td>
      <td>12</td>
      <td>32</td>
      <td>1</td>
      <td>4</td>
      <td></td>
      <td>7</td>
      <td>1</td>
      <td>2</td>
      <td>5</td>
    </tr>
    <tr>
      <th>2020</th>
      <td>38</td>
      <td>3</td>
      <td>4</td>
      <td>2</td>
      <td>4</td>
      <td>1</td>
      <td>2</td>
      <td>1</td>
      <td>3</td>
      <td>4</td>
      <td>4</td>
      <td></td>
    </tr>
  </tbody>
</table>
</div>

> We note an early focus `ibis.sqlite, sqlalchemy and ibis.impala`. 
Later, work began on the `pandas` backend rounding out the three different types of backgrounds.
From this point, improvements were made to these key backends as `ibis.clickhouse, ibis.spark and "postgres"`. Recently, 
For the past 3 years, Quansight, in partnership with OmniSci, added the `ibis.omniscidb`

> string generating backend. Since Quansight Labs has taken on a role as a community maintainer 

Reviewer: [Since Quansight Labs has taken on a role as a community maintainer ] incomplete sentence. I'm not sure how to work this in without downplaying the partnership with Omnisci. But I can see how noting both are important.

for `ibis`. This collaboration introduced geospatial functionality to `ibis`, and we 

> have on going efforts to introduce [SQL Server][sql-server] support

Reviewer: [SQL Server][sql-server] - Is this supposed to be a link?

### Currently, there is an ongoing effort to add sqlserver backends. 

> Leave a comment the backends what you like to see? `dask`? `altair`?

Reviewer: Either fill this out a bit or remove it completely. At the very least, the last sentence from the previous paragraph should be moved to this section.

## `ibis` direct execution.

> `ibis` direct execution backends like `pandas and hdf5` operate on conventional in-memory python objects.

Reviewer: ```suggestion
`ibis` direct execution backends like `pandas` and `hdf5` operate on conventional in-memory python objects.
```

> `pandas` is the gold standard for structured data in python, and inspires the api for `ibis`.

Reviewer: ```suggestion
`pandas` is the gold standard for structured data in python, and inspires the API for `ibis`.
```

> `pd` is an `ibis` backend based off `pandas.DataFrame` objects.

Reviewer: Is `pd` and ibis command? like `ibis.pd`?

Reviewer: this sentence just seems out of place.

Author: this sentence was missing code cause i am too careless in sharing things.

> `expression` is an `ibis` query, that has `expression.compile` and `expression.execute` methods.

Reviewer: Is `expression` and ibis method like `ibis.expression`?

Reviewer: Maybe some leading verbage at the beginning of this sentence like: 'Ibis has the concept of an expression which represents a query`

Author: hopefully this makes more sense with the code.

We'll recognize the __execute__ method when we return `pandas.DataFrame`s from `ibis` expression.
The __compile__ method does not trigger any computation, rather it constructs an intermediate form
that is interpretted by a backend.

In the case of direction execution backends, the `expression` compiles to an the original `ibis` 
expression.  And the computation is carried out based on a set of recipes defined in `ibis`.

In general, we would typically do this work directly in `pandas`, however this work is
practical in mocking tests for expressions independent of backends.

> > Learn more about the [HDF5 direct execution backend in the `ibis` tests][test-hdf5].

Reviewer: add links

Reviewer: still add links

## `ibis` expression generating backends.

Expression generating backends operate on [SQL] databases that interoperator with `sqlalchemy`.

    
In the case of expression generating backends, the intermediate representation is a `sqlalchemy` object.
`sqlalchemy` is _The Database Toolkit for Python_, and `ibis` leverages it compatability
with traditional [SQL] databases.
    

## `ibis` string generating backends.

```bash

> pip install --upgrade ibis-framework[omniscidb]

Reviewer: So I was recently led astray by the ibis docs a few weeks ago. I didn't realize that if you install `ibis-framework` from conda-forge, you get ALL the backends. I'd also add that comment here `conda install -c conda-forge ibis-framework`

```

String generating backends allow `ibis` to interface with big data systems that manage 
their own computation. For example, we may connect to an example `omnisci` database.
    

    
`omnisci` is described as a string generating backend because the intermediate representation of the
query is a flavor of SQL.

    

> A string generating expression compiles to `ibis.omniscidb` flavored [SQL], while `ibis.bigquery` may have a different string representatin.

Reviewer: link?

> A string generating expression compiles to `ibis.omniscidb` flavored [SQL], while `ibis.bigquery` may have a different string representatin.

Reviewer: ```suggestion
A string generating expression compiles to `ibis.omniscidb` flavored [SQL], while `ibis.bigquery` may have a different string representation.
```

>     'SELECT *\nFROM upstream_reservoir\nLIMIT 5'

Reviewer: show me the ibis call that this query string was created from

> Major credit goes to [@xmnlab] in his heroic PR to introduce `ibis.omniscidb` into `ibis`. You can watch

> the drama play out in this [Github Issue][omnisci-pr]. If you'd like to learn more about [OmniSci] and

Reviewer: links

`ibis.omniscidb` checkout the following links.

> > * [OmniSci][omnisci]

Reviewer: links

> > * [Quansight Labs- Ibis: Python data analysis productivity framework][labs-post]

Reviewer: links
